Object files can only be linked with legacy clang linker on Apple

Hello, thank you very much for all of the incredible software y'all are writing. I'm currently using Cranelift to generate native binaries for my project, and I'm continually impressed at the level of detail and effort that goes into these things. Not to mention the super helpful Zulip chat.

In the aforementioned chat, I was receiving help from @bjorn3 and @cfallin on generating executables for an Apple silicon Mac. The following code does not successfully execute on an M-series processor Mac: https://gist.github.com/sezna/e358a9167d4ef3063f6cfe4ce4c2ad37

Using the legacy linker, that is, adding `-Wl -ld_classic` to the clang invocation, _does_ work. However, as the legacy linker will likely be going away someday, I'm filing this issue in the hopes that we can figure out why it doesn't work with the current linker. Note that `cg_clif` also uses `-Wl -ld_classic`, so when the legacy linker is dropped, `cg_clif` could also have issues.

A simple

```rust
#[inline(never)]
pub unsafe fn foo() -> (unsafe extern "C" fn(), &'static u8) {
    bar(); (bar, &FOO)
}

extern "C" {
    fn bar();
    static FOO: u8;
}
```

produces

```
rust_out.o:     file format mach-o arm64

Disassembly of section __TEXT,__text:

0000000000000000 <ltmp0>:
       0: a9bf7bfd      stp     x29, x30, [sp, #-16]!
       4: 910003fd      mov     x29, sp
       8: 94000000      bl      0x8 <ltmp0+0x8>
                0000000000000008:  ARM64_RELOC_BRANCH26 _bar
       c: 90000000      adrp    x0, 0x0 <ltmp0+0xc>
                000000000000000c:  ARM64_RELOC_GOT_LOAD_PAGE21  _bar
      10: f9400000      ldr     x0, [x0]
                0000000000000010:  ARM64_RELOC_GOT_LOAD_PAGEOFF12       _bar
      14: 90000001      adrp    x1, 0x0 <ltmp0+0x14>
                0000000000000014:  ARM64_RELOC_GOT_LOAD_PAGE21  _FOO
      18: f9400021      ldr     x1, [x1]
                0000000000000018:  ARM64_RELOC_GOT_LOAD_PAGEOFF12       _FOO
      1c: a8c17bfd      ldp     x29, x30, [sp], #16
      20: d65f03c0      ret
```

with the LLVM backend. If I try to link it I get a linker error as expected:

```
Undefined symbols for architecture arm64:
  "_FOO", referenced from:
      rust_out::foo::he5a9b1ed9cb33bc0 in rust_out.o
  "_bar", referenced from:
      rust_out::foo::he5a9b1ed9cb33bc0 in rust_out.o
      rust_out::foo::he5a9b1ed9cb33bc0 in rust_out.o
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
```

If I apply the following patch to Cranelift:

```diff
diff --git a/cranelift/codegen/src/isa/aarch64/abi.rs b/cranelift/codegen/src/isa/aarch64/abi.rs
index 922e4ebfc..6d43c1db6 100644
--- a/cranelift/codegen/src/isa/aarch64/abi.rs
+++ b/cranelift/codegen/src/isa/aarch64/abi.rs
@@ -1027,7 +1027,7 @@ impl ABIMachineSpec for AArch64MachineDeps {
     ) -> SmallVec<[Inst; 2]> {
         let mut insts = SmallVec::new();
         match &dest {
-            &CallDest::ExtName(ref name, RelocDistance::Near) => insts.push(Inst::Call {
+            &CallDest::ExtName(ref name, _/*RelocDistance::Near*/) => insts.push(Inst::Call {
                 info: Box::new(CallInfo {
                     dest: name.clone(),
                     uses,
```

Relocations that print identical by `objdump -dr` are generated by Cranelift:

```
rust_out.o:     file format mach-o arm64

Disassembly of section __TEXT,__text:

0000000000000000 <__ZN8rust_out3foo17he5a9b1ed9cb33bc0E>:
       0: a9bf7bfd      stp     x29, x30, [sp, #-16]!
       4: 910003fd      mov     x29, sp
       8: d10043ff      sub     sp, sp, #16
       c: 94000000      bl      0xc <__ZN8rust_out3foo17he5a9b1ed9cb33bc0E+0xc>
                000000000000000c:  ARM64_RELOC_BRANCH26 _bar
      10: 90000000      adrp    x0, 0x0 <__ZN8rust_out3foo17he5a9b1ed9cb33bc0E+0x10>
                0000000000000010:  ARM64_RELOC_GOT_LOAD_PAGE21  _bar
      14: f9400000      ldr     x0, [x0]
                0000000000000014:  ARM64_RELOC_GOT_LOAD_PAGEOFF12       _bar
      18: 90000007      adrp    x7, 0x0 <__ZN8rust_out3foo17he5a9b1ed9cb33bc0E+0x18>
                0000000000000018:  ARM64_RELOC_GOT_LOAD_PAGE21  _FOO
      1c: f94000e7      ldr     x7, [x7]
                000000000000001c:  ARM64_RELOC_GOT_LOAD_PAGEOFF12       _FOO
      20: 910003e8      mov     x8, sp
      24: f9000107      str     x7, [x8]
      28: 910003e8      mov     x8, sp
      2c: f9400101      ldr     x1, [x8]
      30: 910043ff      add     sp, sp, #16
      34: a8c17bfd      ldp     x29, x30, [sp], #16
      38: d65f03c0      ret
```

Yet the object file produced by LLVM gives a regular linker error due to undefined symbols, while the one produced by Cranelift crashes the linker:

```
0  0x100e02074  __assert_rtn + 72
1  0x100d3edb8  ld::InputFiles::SliceParser::parseObjectFile(mach_o::Header const*) const + 22712
2  0x100d4b830  ld::InputFiles::parseAllFiles(void (ld::AtomFile const*) block_pointer)::$_8::operator()(unsigned long, ld::FileInfo const&) const + 440
3  0x189fa6428  _dispatch_client_callout2 + 20
4  0x189fba850  _dispatch_apply_invoke3 + 336
5  0x189fa63e8  _dispatch_client_callout + 20
6  0x189fa7c68  _dispatch_once_callout + 32
7  0x189fbaeec  _dispatch_apply_invoke_and_wait + 372
8  0x189fb9e9c  _dispatch_apply_with_attr_f + 1212
9  0x189fba08c  dispatch_apply + 96
10  0x100dd0564  ld::AtomFileConsolidator::parseFiles(bool) + 292
11  0x100d6bee8  main + 9532
ld: Assertion failed: (pattern[0].addrMode == addr_other), function addFixupFromRelocations, file Relocations.cpp, line 701.
```

This means there has to be a difference somewhere. If I run bingrep on both binaries, the relocations are shown in opposite order for both executable. For LLVM:

```
Relocations(5):
  Segment   Section   Count  
  __TEXT    __text    5      
         Type                             Offset   Length   PIC      Extern   SymbolNum   Symbol  
         ARM64_RELOC_GOT_LOAD_PAGEOFF12   0x18     2        false    true     0x2         _FOO    
         ARM64_RELOC_GOT_LOAD_PAGE21      0x14     2        true     true     0x2         _FOO    
         ARM64_RELOC_GOT_LOAD_PAGEOFF12   0x10     2        false    true     0x3         _bar    
         ARM64_RELOC_GOT_LOAD_PAGE21      0xc      2        true     true     0x3         _bar    
         ARM64_RELOC_BRANCH26             0x8      2        true     true     0x3         _bar    
```

and for Cranelift:

```
Relocations(5):
  Segment   Section   Count  
  __TEXT    __text    5      
         Type                             Offset   Length   PIC      Extern   SymbolNum   Symbol  
         ARM64_RELOC_BRANCH26             0xc      2        true     true     0x2         _bar    
         ARM64_RELOC_GOT_LOAD_PAGE21      0x10     2        true     true     0x2         _bar    
         ARM64_RELOC_GOT_LOAD_PAGEOFF12   0x14     2        false    true     0x2         _bar    
         ARM64_RELOC_GOT_LOAD_PAGE21      0x18     2        true     true     0x1         _FOO    
         ARM64_RELOC_GOT_LOAD_PAGEOFF12   0x1c     2        false    true     0x1         _FOO    
```

As hack I tried to reverse the order of relocations emitted by Cranelift too, which worked.

```diff
diff --git a/cranelift/codegen/src/isa/aarch64/abi.rs b/cranelift/codegen/src/isa/aarch64/abi.rs
index 922e4ebfc..6d43c1db6 100644
--- a/cranelift/codegen/src/isa/aarch64/abi.rs
+++ b/cranelift/codegen/src/isa/aarch64/abi.rs
@@ -1027,7 +1027,7 @@ impl ABIMachineSpec for AArch64MachineDeps {
     ) -> SmallVec<[Inst; 2]> {
         let mut insts = SmallVec::new();
         match &dest {
-            &CallDest::ExtName(ref name, RelocDistance::Near) => insts.push(Inst::Call {
+            &CallDest::ExtName(ref name, _/*RelocDistance::Near*/) => insts.push(Inst::Call {
                 info: Box::new(CallInfo {
                     dest: name.clone(),
                     uses,
diff --git a/cranelift/object/src/backend.rs b/cranelift/object/src/backend.rs
index 2d48e9b61..c5a84f14c 100644
--- a/cranelift/object/src/backend.rs
+++ b/cranelift/object/src/backend.rs
@@ -501,7 +501,7 @@ impl ObjectModule {
                 ref name,
                 flags,
                 addend,
-            } in &symbol.relocs
+            } in symbol.relocs.iter().rev()
             {
                 let target_symbol = self.get_symbol(name);
                 self.object
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -336,7 +336,7 @@ regalloc2 = "0.15.1"
 wasip1 = { version = "1.0.0", default-features = false }
 
 # cap-std family:
-target-lexicon = "0.13.0"
+target-lexicon = "0.13.3"
 cap-std = "3.4.5"
 cap-fs-ext = "3.4.5"
 cap-net-ext = "3.4.5"

--- a/cranelift/codegen/src/isa/aarch64/mod.rs
+++ b/cranelift/codegen/src/isa/aarch64/mod.rs
@@ -18,8 +18,6 @@ use alloc::string::String;
 use alloc::{boxed::Box, vec::Vec};
 use core::fmt;
 use cranelift_control::ControlPlane;
-#[cfg(feature = "unwind")]
-use target_lexicon::OperatingSystem;
 use target_lexicon::{Aarch64Architecture, Architecture, Triple};
 
 // New backend:
@@ -151,17 +149,9 @@ impl TargetIsa for AArch64Backend {
 
     #[cfg(feature = "unwind")]
     fn create_systemv_cie(&self) -> Option<gimli::write::CommonInformationEntry> {
-        let is_apple_os = match self.triple.operating_system {
-            OperatingSystem::Darwin(_)
-            | OperatingSystem::IOS(_)
-            | OperatingSystem::MacOSX { .. }
-            | OperatingSystem::TvOS(_) => true,
-            _ => false,
-        };
-
         if self.isa_flags.sign_return_address()
             && self.isa_flags.sign_return_address_with_bkey()
-            && !is_apple_os
+            && !self.triple.operating_system.is_like_darwin()
         {
             unimplemented!(
                 "Specifying that the B key is used with pointer authentication instructions in the CIE is not implemented."
@@ -185,19 +175,12 @@ impl TargetIsa for AArch64Backend {
     }
 
     fn page_size_align_log2(&self) -> u8 {
-        use target_lexicon::*;
-        match self.triple().operating_system {
-            OperatingSystem::MacOSX { .. }
-            | OperatingSystem::Darwin(_)
-            | OperatingSystem::IOS(_)
-            | OperatingSystem::TvOS(_) => {
-                debug_assert_eq!(1 << 14, 0x4000);
-                14
-            }
-            _ => {
-                debug_assert_eq!(1 << 16, 0x10000);
-                16
-            }
+        if self.triple().operating_system.is_like_darwin() {
+            debug_assert_eq!(1 << 14, 0x4000);
+            14
+        } else {
+            debug_assert_eq!(1 << 16, 0x10000);
+            16
         }
     }
 

--- a/cranelift/object/src/backend.rs
+++ b/cranelift/object/src/backend.rs
@@ -10,7 +10,7 @@ use cranelift_module::{
     DataDescription, DataId, FuncId, Init, Linkage, Module, ModuleDeclarations, ModuleError,
     ModuleReloc, ModuleRelocTarget, ModuleResult,
 };
-use log::info;
+use log::{info, warn};
 use object::write::{
     Object, Relocation, SectionId, StandardSection, Symbol, SymbolId, SymbolSection,
 };
@@ -21,7 +21,7 @@ use object::{
 use std::collections::HashMap;
 use std::collections::hash_map::Entry;
 use std::mem;
-use target_lexicon::PointerWidth;
+use target_lexicon::{PointerWidth, Triple};
 
 /// A builder for `ObjectModule`.
 pub struct ObjectBuilder {
@@ -137,6 +137,68 @@ impl ObjectBuilder {
     }
 }
 
+/// See the following for details:
+/// <https://github.com/rust-lang/rust/blob/1.95.0/compiler/rustc_codegen_ssa/src/back/metadata.rs#L408-L425>
+fn macho_build_version(triple: &Triple) -> Option<object::write::MachOBuildVersion> {
+    use target_lexicon::{DeploymentTarget, OperatingSystem::*};
+
+    fn pack_version(v: DeploymentTarget) -> u32 {
+        let (major, minor, patch) = (v.major as u32, v.minor as u32, v.patch as u32);
+        (major << 16) | (minor << 8) | patch
+    }
+
+    match triple.operating_system {
+        Darwin(v) | MacOSX(v) | IOS(v) | TvOS(v) | VisionOS(v) | WatchOS(v) | XROS(v) => {
+            use object::macho::*;
+            use target_lexicon::Environment::*;
+            // Same as https://github.com/rust-lang/rust/blob/1.95.0/compiler/rustc_codegen_ssa/src/back/apple.rs#L36-L50.
+            //
+            // TODO(madsmtm): Properly support simulator after
+            // https://github.com/bytecodealliance/target-lexicon/pull/130
+            let platform = match (triple.operating_system, triple.environment) {
+                (Darwin(_), _) => 0, // PLATFORM_UNKNOWN
+                (MacOSX(_), _) => PLATFORM_MACOS,
+                (_, Macabi) => PLATFORM_MACCATALYST,
+                (IOS(_), Sim) => PLATFORM_IOSSIMULATOR,
+                (IOS(_), _) => PLATFORM_IOS,
+                (TvOS(_), Sim) => PLATFORM_TVOSSIMULATOR,
+                (TvOS(_), _) => PLATFORM_TVOS,
+                (VisionOS(_) | XROS(_), Sim) => PLATFORM_XROSSIMULATOR,
+                (VisionOS(_) | XROS(_), _) => PLATFORM_XROS,
+                (WatchOS(_), Sim) => PLATFORM_WATCHOSSIMULATOR,
+                (WatchOS(_), _) => PLATFORM_WATCHOS,
+                _ => {
+                    warn!("unsupported OS/environment: {triple}");
+                    0
+                }
+            };
+
+            let mut build_version = object::write::MachOBuildVersion::default();
+            build_version.platform = platform;
+
+            build_version.minos = if let Some(v) = v {
+                pack_version(v)
+            } else {
+                // The `minos` in object files is useful for diagnostics, as
+                // it tells the linker whether the file supports a given OS -
+                // if the `minos` is higher than what you're linking against,
+                // that's a signal that something has gone wrong.
+                //
+                // Using `0.0.0` here should be fine if we don't have the data
+                // available.
+                0
+            };
+
+            // Setting a 0 SDK version is fine, it's only relevant for the
+            // final linked binary.
+            build_version.sdk = 0;
+
+            Some(build_version)
+        }
+        _ => None,
+    }
+}
+
 /// An `ObjectModule` implements `Module` and emits ".o" files using the `object` library.
 ///
 /// See the `ObjectBuilder` for a convenient way to construct `ObjectModule` instances.
@@ -162,6 +224,13 @@ impl ObjectModule {
         object.flags = builder.flags;
         object.set_subsections_via_symbols();
         object.add_file_symbol(builder.name);
+        if let Some(info) = macho_build_version(builder.isa.triple()) {
+            // Set LC_BUILD_VERSION.
+            //
+            // Required when linking Apple targets to avoid warning, see:
+            // https://github.com/bytecodealliance/wasmtime/issues/8730
+            object.set_macho_build_version(info);
+        }
         Self {
             isa: builder.isa,
             object,

--- a/supply-chain/imports.lock
+++ b/supply-chain/imports.lock
@@ -1037,8 +1037,8 @@ user-login = "alexcrichton"
 user-name = "Alex Crichton"
 
 [[publisher.target-lexicon]]
-version = "0.13.0"
-when = "2024-12-05"
+version = "0.13.3"
+when = "2025-09-09"
 user-id = 6825
 user-login = "sunfishcode"
 user-name = "Dan Gohman"

